File: signal_gen.py
# tools/signal_gen.py
import sys
import math
import wave
import struct

PI = 3.141592653589793

def s_j(t: float, f: float, j: int) -> float:
    """Same 4 waveforms as the C version."""
    if j == 0:
        # sine
        return math.sin(2 * PI * f * t)
    elif j == 1:
        # sawtooth: f*t - floor(f*t)
        return (f * t) - math.floor(f * t)
    elif j == 2:
        # square: sign(sin)
        return 1.0 if math.sin(2 * PI * f * t) >= 0 else -1.0
    elif j == 3:
        # triangle: 2*fabs(2*(f*t - floor(f*t+0.5))) - 1
        return 2.0 * abs(2.0 * (f * t - math.floor(f * t + 0.5))) - 1.0
    else:
        return 0.0

def clamp_int16(x: float) -> int:
    if x > 32767:
        return 32767
    if x < -32768:
        return -32768
    return int(x)

def main():
    # Keep the same CLI as C:
    # Usage: signal_gen.py sample_rate output.wav
    if len(sys.argv) < 3:
        print(f"Usage: {sys.argv[0]} sample_rate output.wav")
        sys.exit(1)

    fs = int(sys.argv[1])
    out_wav = sys.argv[2]

    duration = 4
    samples = fs * duration

    a = [100, 2000, 1000, 500, 250, 100, 2000, 1000, 500, 250]
    f = [0, 31.25, 500, 2000, 4000, 44, 220, 440, 1760, 3960]

    # Write WAV: mono, 16-bit PCM
    with wave.open(out_wav, "wb") as w:
        w.setnchannels(1)
        w.setsampwidth(2)     # 16-bit
        w.setframerate(fs)

        frames = bytearray()

        for n in range(samples):
            t = n / fs
            x = 0.0

            for j in range(4):
                for i in range(10):
                    start = 0.1 * i + j
                    end = 0.1 * (i + 1) + j
                    if start <= t < end:
                        x += a[i] * s_j(t - start, f[i], j)

            sample = clamp_int16(x)
            frames += struct.pack("<h", sample)

        w.writeframes(frames)

if __name__ == "__main__":
    main()

File: signal_gen.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <math.h>

#define PI 3.141592653589793

// 四種波形
double s_j(double t, double f, int j) {
    switch(j) {
        case 0: return sin(2*PI*f*t); // sine
        case 1: return f*t - floor(f*t); // sawtooth
        case 2: return (sin(2*PI*f*t) >= 0 ? 1.0 : -1.0); // square
        case 3: return 2*fabs(2*(f*t - floor(f*t+0.5))) - 1; // triangle
        default: return 0.0;
    }
}

// 簡單 WAV header
void write_wav_header(FILE *fp, int sample_rate, int num_samples) {
    int byte_rate = sample_rate * 2;
    int block_align = 2;
    int subchunk2_size = num_samples * 2;
    int chunk_size = 36 + subchunk2_size;

    fwrite("RIFF",1,4,fp);
    fwrite(&chunk_size,4,1,fp);
    fwrite("WAVEfmt ",1,8,fp);
    int subchunk1_size=16; short audio_format=1; short num_channels=1;
    fwrite(&subchunk1_size,4,1,fp);
    fwrite(&audio_format,2,1,fp);
    fwrite(&num_channels,2,1,fp);
    fwrite(&sample_rate,4,1,fp);
    fwrite(&byte_rate,4,1,fp);
    fwrite(&block_align,2,1,fp);
    short bits_per_sample=16;
    fwrite(&bits_per_sample,2,1,fp);
    fwrite("data",1,4,fp);
    fwrite(&subchunk2_size,4,1,fp);
}

int main(int argc, char *argv[]) {
    if(argc<3) {
        printf("Usage: %s sample_rate output.wav\n",argv[0]);
        return -1;
    }
    int fs=atoi(argv[1]);
    int duration=4;
    int samples=fs*duration;
    FILE *fp=fopen(argv[2],"wb");
    write_wav_header(fp,fs,samples);

    int a[10]={100,2000,1000,500,250,100,2000,1000,500,250};
    double f[10]={0,31.25,500,2000,4000,44,220,440,1760,3960};

    for(int n=0;n<samples;n++){
        double t=(double)n/fs;
        double x=0.0;
        for(int j=0;j<4;j++){
            for(int i=0;i<10;i++){
                double start=0.1*i+j;
                double end=0.1*(i+1)+j;
                if(t>=start && t<end){
                    x+=a[i]*s_j(t-start,f[i],j);
                }
            }
        }
        if(x>32767) x=32767;
        if(x<-32768) x=-32768;
        int16_t sample=(int16_t)x;
        fwrite(&sample,2,1,fp);
    }
    fclose(fp);
    return 0;
}